File: 0064-minimum-path-sum/0064-minimum-path-sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        def findpath(i, j):
            nonlocal m,n
            if i==0 and j==0:
                return grid[i][j]
            elif 0<=i<m and 0<=j<n:
                if memo[i][j]==-1:
                    memo[i][j] = min(findpath(i-1, j), findpath(i, j-1)) + grid[i][j]
                return memo[i][j]
            return 10**7
        
        m,n = len(grid), len(grid[0])
        memo = [[-1]*n for _ in range(m)]
        return findpath(m-1, n-1)

File: 0064-minimum-path-sum/0064-minimum-path-sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        return (int)(travel(grid, 0,0,new int[grid.length][grid[0].length]));
    }
    double travel(int[][] grid, int i, int j, int[][] map){
        if(i<0 || i>=grid.length || j<0 || j>=grid[0].length) return Integer.MAX_VALUE;
        if(i==grid.length-1 && j==grid[0].length-1) return grid[i][j];
        if(map[i][j]==0) map[i][j]=(int)(Math.min(travel(grid, i+1, j, map),travel(grid, i, j+1, map)))+grid[i][j];
        return map[i][j];
    }
}


